### PR TITLE
Support large batch tensors in grouped conv bwd data

### PR DESCRIPTION
**Pull request by bartekxk**

### CODE REVIEW

File: include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_data_multiple_d_wmma_cshuffle.hpp [hunk 1]
```diff
@@ -106,7 +106,7 @@ struct DeviceGroupedConvBwdDataMultipleD_Wmma_CShuffle
     static constexpr auto I3           = Number<3>{};
     static constexpr index_t KPerBlock = K0PerBlock * K1;
 
-    static constexpr auto transform_conv_to_gemm =
+    using ConvToGemmBwdDataTransformer =
```
Comment (geyyer) [inline include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_data_multiple_d_wmma_cshuffle.hpp]: Could you rephrase this alias? I understand what you mean, but "Transformer" looks a little misleading.

File: include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_data_multiple_d_xdl_cshuffle_v1.hpp [hunk 2]
```diff
@@ -526,68 +473,66 @@ struct DeviceGroupedConvBwdDataMultipleD_Xdl_CShuffle_v1
...
+                                                              NPerBlock,
+                                                              KPerBlock,
+                                                              DoPadGemmM,
+                                                              DoPadGemmN,
+                                                              ALayout,
+                                                              BLayout,
+                                                              ELayout,
+                                                              true, /*SplitConvN*/
+                                                              ABDataType,
+                                                              DLayout>;
```
Comment (geyyer) [inline include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_data_multiple_d_xdl_cshuffle_v1.hpp]: Would it make sense to squeeze DLayout in the middle of other Layout arguments?

File: include/ck/tensor_operation/operator_transform/transform_conv_bwd_data_to_gemm_v1.hpp [hunk 3]
```diff
@@ -13,245 +13,613 @@
...
+        HTilde_ = Ho_ + math::integer_divide_ceil(ConvDilationH_ * (Y_ - I1), ConvStrideH_);
+        WTilde_ = Wo_ + math::integer_divide_ceil(ConvDilationW_ * (X_ - I1), ConvStrideW_);
+
+        YDot_ = math::integer_divide_ceil(Y_, YTilde_);
+        XDot_ = math::integer_divide_ceil(X_, XTilde_);
     }
-    else if constexpr(is_same_v<CLayout, tensor_layout::convolution::GNDHWC> ||
-                      is_same_v<CLayout, tensor_layout::convolution::NDHWGC>)
+
+#if 0 // At now not supported to split tensor
```
Comment (geyyer) [inline include/ck/tensor_operation/operator_transform/transform_conv_bwd_data_to_gemm_v1.hpp:285]: Do we need this chunk?

File: include/ck/tensor_operation/operator_transform/transform_conv_bwd_data_to_gemm_v1.hpp [hunk 4]
```diff
@@ -482,66 +833,31 @@ struct TransformConvBwdDataToGemm_v1
         }
     }
 
-    template <typename BLayout,
+    template <typename BLay                       = BLayout,
```
Comment (geyyer) [inline include/ck/tensor_operation/operator_transform/transform_conv_bwd_data_to_gemm_v1.hpp]: Maybe something like BLayout_ ?

**Review by geyyer — DISMISSED** [review]
LGTM, just a couple of remarks!

File: include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_data_multiple_d_xdl_cshuffle_v1.hpp (hunk 2 above)
Comment (bartekxk) [reply include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_data_multiple_d_xdl_cshuffle_v1.hpp]: I think no because we use the same MakeDescriptor for C and D.

File: include/ck/tensor_operation/operator_transform/transform_conv_bwd_data_to_gemm_v1.hpp (hunk 3 above)
Comment (bartekxk) [reply include/ck/tensor_operation/operator_transform/transform_conv_bwd_data_to_gemm_v1.hpp:285]: Generally it is some basic implementation of support image larger than 2GB. So I think it is worth to keep this.

File: test/grouped_convnd_bwd_data/test_grouped_convnd_bwd_data_wmma.cpp [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+#include <initializer_list>
+#include <tuple>
+#include <vector>
+
+#include <gtest/gtest.h>
+
+#include "profiler/profile_grouped_conv_bwd_data_impl.hpp"
+
+template <typename Tuple>
+class TestGroupedConvndBwdData : public ::testing::Test
```
Comment (aosewski) [inline test/grouped_convnd_bwd_data/test_grouped_convnd_bwd_data_wmma.cpp]: Maybe we could add suffix `Wmma` in order to differentiate it?

File: test/grouped_convnd_bwd_data/test_grouped_convnd_bwd_data_wmma.cpp [hunk 6]
```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2024, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (aosewski) [inline test/grouped_convnd_bwd_data/test_grouped_convnd_bwd_data_wmma.cpp]:
```suggestion
// Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
```

File: test/grouped_convnd_bwd_data/test_grouped_convnd_bwd_data_xdl.cpp
Comment (aosewski) [inline test/grouped_convnd_bwd_data/test_grouped_convnd_bwd_data_xdl.cpp:1]: Maybe we could add suffix `Xdl` to test class in order to differentiate it?

File: include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_data_multiple_d_wmma_cshuffle.hpp [hunk 7]
```diff
@@ -106,89 +106,39 @@ struct DeviceGroupedConvBwdDataMultipleD_Wmma_CShuffle
...
-                                                                         dummy_spatial_lengths,
-                                                                         dummy_spatial_lengths);
+        const auto a_grid_desc_ak0_m_ak1 = conv_to_gemm_transform.MakeADescriptor_AK0_M_AK1();
+
+        const auto b_grid_desc_bk0_n_bk1 = conv_to_gemm_transform.MakeBDescriptor_BK0_N_BK1();
+
+        const auto ds_grid_desc_m_n =
+            generate_tuple([&](auto) { return conv_to_gemm_transform.MakeCDescriptor_M_N(); },
+                           Number<NumDTensor>{});
+
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_data_multiple_d_wmma_cshuffle.hpp]: Please remove all those empty lines.

File: include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_data_multiple_d_xdl_cshuffle_v1.hpp (hunk 2 above)
Comment (aosewski) [reply include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_data_multiple_d_xdl_cshuffle_v1.hpp]: I think Rosty had in mind reordering of arguments such that DLayout could be ie right after `BLayout` ? I think this is a good idea. It doesn't matter that both E and D could have same layout.

File: include/ck/tensor_operation/operator_transform/transform_conv_bwd_data_to_gemm_v1.hpp [hunk 8]
```diff
@@ -13,245 +13,614 @@
...
-                     ck::tensor_operation::device::ConvolutionBackwardDataSpecialization::
-                         Filter1x1Stride1Pad0)
-        {
+    static constexpr auto NonSpatialDimsNum = Number<3>{};
 
-            return make_naive_tensor_descriptor(make_tuple(N * Ho * Wo, K),
-                                                make_tuple(WiStride, KStride));
-        }
-        else
+    static constexpr auto DIdx = Number<NonSpatialDimsNum>{};
```
Comment (aosewski) [inline include/ck/tensor_operation/operator_transform/transform_conv_bwd_data_to_gemm_v1.hpp]: Why Number of Number? Doesn't it work like that: 
```
static constexpr auto DIdx = NonSpatialDimsNum;

File: include/ck/tensor_operation/operator_transform/transform_conv_bwd_data_to_gemm_v1.hpp [hunk 9]
```diff
@@ -13,245 +13,614 @@
...
+            // Split N is not needed.
+            return N;
         }
     }
-    else if constexpr(is_same_v<ALayout, tensor_layout::convolution::GNHWK>)
+
+    public:
+    __host__ __device__ constexpr TransformConvBwdDataToGemm_v1() {}
+
+    template <typename TransformConvBwdDataToGemm_v1Base>
```
Comment (aosewski) [inline include/ck/tensor_operation/operator_transform/transform_conv_bwd_data_to_gemm_v1.hpp:124]: What is this base? If you just want copy constructor then `TransformConvBwdDataToGemm_v1` should suffice.

**Review by aosewski — COMMENTED** [review]
Looks great! Few minor comments.

File: include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_data_multiple_d_xdl_cshuffle_v1.hpp (hunk 2 above)
Comment (geyyer) [reply include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_data_multiple_d_xdl_cshuffle_v1.hpp]: Exactly, I was thinking about the argument order.

Comment (bartekxk) [reply include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_data_multiple_d_xdl_cshuffle_v1.hpp]: It was a bug. Fixed.

File: include/ck/tensor_operation/operator_transform/transform_conv_bwd_data_to_gemm_v1.hpp (hunk 9 above)
Comment (bartekxk) [reply include/ck/tensor_operation/operator_transform/transform_conv_bwd_data_to_gemm_v1.hpp:124]: Because index type could be different. (template argument of TransformConvBwdDataToGemm structure)

**Review by aosewski — APPROVED** [review]
